src/closure.md

```diff
@@ -0,0 +1,113 @@
+# Closure Expansion in rustc
+
+Let's start with an example
+
+### Example 1
+```rust
+fn foo<F>(closure: F)
+  where F: Fn() {
+  closure()
+}
+fn main() {
+  let x = 10;
+  foo(|| println!("Hi {}", x))
+}
+```
+If you compile Example 1 using the command
+```
+rustc +stage1 test.rs -Zdump-mir=all
+```
+where the name of the file is test.rs. This command will generate a directory called mir_dump
+which will contain several files. If we look at file `rustc.main.-------.mir_map.0.mir`, we will find this
+
+```rust,ignore
+  _3 = [closure@test.rs:9:9: 11:6] { x: move _4 }; // bb0[6]: scope 1 at test.rs:9:9: 11:6
+                                   // closure
+                                   // + def_id: DefId(0/1:10 ~ test[317d]::main[0]::{{closure}}[0])
+                                   // + substs: ClosureSubsts {
+                                   //     substs: [
+                                   //         i8,
+                                   //         extern "rust-call" fn(()),
+                                   //         &i32
+                                   //     ]
+                                   // }
+```
```

Comment:
Here, I think we should say a bit more =)

This is showing the construction of the desugared closure struct. The `[closure@test.rs...]` type is the "name" of the closure struct, and the `{ x : move _4 }` shows the value for each of its fields. In this case, we can see that we are *moving* the variable `_4` into the closure struct. If we look back to see where `_4` is defined ...

(I left the last sentence finished: what do we find if we do that? I presume it is a borrow of `x`...